backends/tc/backend.cpp

```diff
@@ -0,0 +1,623 @@
+                auto paramType = typeMap->getType(param);
+                IR::TCActionParam *tcActionParam = new IR::TCActionParam();
+                tcActionParam->setParamName(param->name.originalName);
+                if (!paramType->is<IR::Type_Bits>()) {
+                    ::error(ErrorType::ERR_UNSUPPORTED_ON_TARGET,
+                            "%1% parameter with type other than bit is not supported", param);
+                    return;
+                } else {
+                    auto paramTypeName = paramType->to<IR::Type_Bits>()->baseName();
+                    if (paramTypeName != "bit") {
+                        ::error(ErrorType::ERR_UNSUPPORTED_ON_TARGET,
+                                "%1% parameter with type other than bit is not supported", param);
+                        return;
+                    }
+                    tcActionParam->setDataType(BIT_TYPE);
+                    unsigned int width = paramType->to<IR::Type_Bits>()->width_bits();
+                    tcActionParam->setBitSize(width);
+                }
+                auto annoList = param->getAnnotations()->annotations;
+                for (auto anno : annoList) {
+                    if (anno->name != "tc_type") continue;
+                    auto expressionList = anno->expr;
+                    for (auto expr : expressionList) {
+                        if (auto typeLiteral = expr->to<IR::StringLiteral>()) {
+                            auto val = checkTcType(typeLiteral);
+                            if (val != BIT_TYPE) {
+                                tcActionParam->setDataType(val);
+                            } else {
+                                ::error(ErrorType::ERR_INVALID,
+                                        "tc_type annotation cannot have '%1%' as value", expr);
+                            }
+                        } else {
+                            ::error(ErrorType::ERR_INVALID,
+                                    "tc_type annotation cannot have '%1%' as value", expr);
+                        }
+                    }
+                }
+                tcAction->addActionParams(tcActionParam);
+            }
+        }
+        tcPipeline->addActionDefinition(tcAction);
+    }
+}
+
+void ConvertToBackendIR::updateDefaultMissAction(const IR::P4Table *t, IR::TCTable *tabledef) {
+    auto defaultAction = t->getDefaultAction();
+    if (defaultAction == nullptr || !defaultAction->is<IR::MethodCallExpression>()) return;
+    auto methodexp = defaultAction->to<IR::MethodCallExpression>();
+    if ((methodexp == nullptr) || (methodexp->method == nullptr) ||
+        (!methodexp->method->is<IR::PathExpression>()))
+        return;
+    auto pathexp = methodexp->method->to<IR::PathExpression>();
+    if (pathexp == nullptr || pathexp->path == nullptr) return;
+    auto actionDecl = refMap->getDeclaration(pathexp->path);
+    auto actionName = externalName(actionDecl);
```

Comment:
I think this could be simplified by using MethodInfo

Reply:
updated the code